### ISSUE TITLE
Lookalike Model, tfrecord to pickle processing script run time = 30 hours

### ISSUE THREAD

**Issue by sanjaynirmal** [Futurewei-io/blue-marlin#161]
This week we were doing a final round of testing of Lookalike Model on big data (number of ucdocs = 277864091, 45 days of show log and click log data)

We ran trainer/lookalike_build_dataset.py which is running quite slow i.e. 30 hours of runtime.

Can we reduce the run time of this script ?

**Comment by sanjaynirmal** [comment]
@jimmylao 
We have solved this issue by parallel processing, now we have pickle file of size 17 GB. So when we run the model training script we get memory overflow error. We tried with the best machine available on MTP but not able to train the model.
Any suggestion on tackling this issue.

**Comment by sanjaynirmal** [comment]
@jimmylao 
We have now changed the training script as per the approach 1 that you shared in the ppt. 
We tried to load the 1st pickle file and run the training it ran fine, we saw the epoch values on the screen.

![image](https://user-images.githubusercontent.com/40193781/120601693-e81fb600-c467-11eb-9d88-f2d3cef1a9e4.png)
But when we entered the second loop, we loaded the 2nd pickle file and in the training part we are encountering an error about reusing the tf variable.

![image](https://user-images.githubusercontent.com/40193781/120601273-5b74f800-c467-11eb-8e5a-b104fce589a0.png)

I tried to set the tf variable reuse=True but then get another error saying that no parameter named "reuse" exists.

**Comment by zqlao** [comment]
@sanjaynirmal  Please have a reference of the attached ppt document (slides 8 & 9) for detailed instruction. Please check the modified code (tested and working) in the attachment and compare with your modification to see if it can solve your problem or not.
[training-sample-batch-generation.pdf](https://github.com/Futurewei-io/blue-marlin/files/6591023/training-sample-batch-generation.pdf)
[train_multiple_pickle.txt](https://github.com/Futurewei-io/blue-marlin/files/6591025/train_multiple_pickle.txt)
[lookalike_build_dataset_multiple_pickle.txt](https://github.com/Futurewei-io/blue-marlin/files/6591029/lookalike_build_dataset_multiple_pickle.txt)

**Comment by sanjaynirmal** [comment]
@zqlao 
I modified my code taking cues from the information that you provided and also did a small testing by preparing a small data using 10 tfrecord files which created 4 pickle files which went into model training part. On this data model training was fine. 

But when I tried to run the code on full data, I am getting a GPU memory overflow error which is due to data large data size which cannot be loaded into the RAM during the training process. 

On some deep analysis I found that the issue is due to a variable names user_count which goes into the tf.model()
![image](https://user-images.githubusercontent.com/40193781/120830078-c1eb3a80-c57b-11eb-88fb-3130b29b0502.png)
with full data user_count = 50,659,958 
If I reduce this number just for testing purpose and say I run for user_count = 5,065,995 then the code runs. 
Is there any way in which we can tackle this huge value during model training.

**Comment by jimmylao** [comment]
@sanjaynirmal 
Please comment out line 40 in model.py as shown below

user_emb_w = tf.get_variable("user_emb_w", [user_count, hidden_units])

It should solve the "GPU memory overflow" issue.

**Comment by jimmylao** [comment]
@sanjaynirmal 
Please comment out line 40 in model.py as shown below

user_emb_w = tf.get_variable("user_emb_w", [user_count, hidden_units])

It should solve the "GPU memory overflow" issue.

**Comment by sanjaynirmal** [comment]
@zqlao 
I am able to run the code now, training is running but there are some things which we need to address.

1. By commenting "user_emb_w" we are actually removing the user embedding vector. So can it have any negative effect on the model learning.
2. The current training code is running sequentially, it is not able to use multiple GPUs (this could be because of the way we are controlling data in every epoch). Expected training time is 4-5 days.

Can we have a meeting today and discuss these things.

**Comment by sanjaynirmal** [comment]
Meeting Title : 
Enter a meeting title
Time : 
2021-06-07 18:30-19:30 (UTC+05:30)Mumbai
Join (Host) : 
http://imeeting.huawei.com/meeting/startzoom?id=8c8b83f97922081a0179e5f27dd51e4f&app=welink
Join (Internal) : 
http://imeeting.huawei.com/meeting/joinzoom?id=947226653&app=welink
Join (External) : 
https://welink-meeting.zoom.us/j/947226653
ID Number : 
947 226 653    (The conference ID is Unavailable after the conference ends)

**Comment by sanjaynirmal** [comment]
Join (External) : 
https://welink-meeting.zoom.us/j/947226653

**Comment by jimmylao** [comment]
Please find the information regarding using "user history" to represent user instead of user id as follows
[https://github.com/zhougr1993/DeepInterestNetwork/issues/6](url)
[https://github.com/zhougr1993/DeepInterestNetwork/issues/3](url)
[https://github.com/zhougr1993/DeepInterestNetwork/issues/64](url)
Since some of the comments are in Chinese, you may use google translate to understand it. The conclusion from the author is:

for large scale data, user characteristic embedding will either cause over-fitting or useless

**Comment by sanjaynirmal** [comment]
@jimmylao 
Thanks for the links, I have also commented the user embedding vector and currently model is training. 

I wanted to know, Does the current training code utilizes all the GPUs of the machine ? 
I am running training on MTP platform with multiple GPUs but the code is not able to utilize multiple GPUs, If we can make use of all available GPUs then our training time will reduce a lot.

**Comment by jimmylao** [comment]
@sanjaynirmal Current training code does not explicitly support multi-GPU training, will need to add multi-GPU support in training code to speed up the training process.

**Comment by sanjaynirmal** [comment]
@jimmylao 
Can we have a short discussion on multi-GPU training approach that you are planning. I will be sharing the meeting link below.

**Comment by sanjaynirmal** [comment]
https://meeting.zhumu.me/j/131226424

**Comment by jimmylao** [comment]
@sanjaynirmal Please resend meeting link, the link above does not work

**Comment by rangaswamymr** [comment]
@jimmylao  just shared new link:
https://meeting.zhumu.me/j/207085920?status=success

**Comment by sanjaynirmal** [comment]
@jimmylao
auc values for 2 epoch
[auc.xlsx](https://github.com/Futurewei-io/blue-marlin/files/6664478/auc.xlsx)

**Comment by jimmylao** [comment]
@sanjaynirmal Can you double check the attached file (auc.xlsx)? Does it have the result of full set data (1600 pickle files)?

**Comment by sanjaynirmal** [comment]
@jimmylao 
You can see the attached log file here. These are raw log of my code from the starting which is run over 1600 pickle files for 2 times. Every time I open a pickle file, I print data length of train_set, test_set and gdin_label.
[raw_log.log](https://github.com/Futurewei-io/blue-marlin/files/6676047/raw_log.log)

**Comment by sanjaynirmal** [comment]
@jimmylao 
We have solved the model training issue.
We have rewritten the data pipeline and now save tfrecords instead of pickle files then these tfrecords are directly used for training. Overall this process has improved the run time. 
Recently we trained the model on 45 days of data which had close to 1.35 billon tfrecords and within 3-4 hours we were able to see the model converge on a single GPU machine. 
Model error in the beginning was around 0.7 and after 3 hours it came to 0.2
